### PR TITLE
16 inference pipeline for baskerville

### PR DESCRIPTION
**Pull request by J-Dymond**
This branch adds:
- Single component version of the pipeline
- Inference scripts which loop over entire dataset and save confidence measures and errors for all inputs
- Tests for the pipelines to ensure correct inputs are being passed to each pipeline stage
- Some utility functions for inference

### CODE REVIEW

File: config/RTC_configs/bert-mt5-bert.yaml
Comment (lannelin) [inline config/RTC_configs/bert-mt5-bert.yaml:1]: the first step of this isn't bert and i'd suggest the last one is better identified by being zero-shot than it being a bert derivative.

File: scripts/pipeline_inference.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    pipeline_name = args.pipeline_config.split("/")[-1].split(".")[0]
+    save_loc = f"{OUTPUT_DIR}/inference_results/{data_name}/{pipeline_name}"
+    os.makedirs(save_loc, exist_ok=True)
+
+    with open(f"{save_loc}/full_pipeline.json", "w") as save_file:
+        json.dump(test_results, save_file)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
```
Comment (lannelin) [inline scripts/pipeline_inference.py:42]: use jsonargparse for consistency with other scripts

File: scripts/single_component_inference.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        f"single_component"
+    )
+    os.makedirs(save_loc, exist_ok=True)
+
+    with open(f"{save_loc}/{args.model_key}.json", "w") as save_file:
+        json.dump(test_results, save_file)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
```
Comment (lannelin) [inline scripts/single_component_inference.py:58]: as prev, use jsonargparse for consistency

File: src/arc_spice/eval/inference_utils.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+        return results_dict
+
+
+def run_inference(
+    dataloader: DataLoader,
+    pipeline: RTCVariationalPipeline | RTCSingleComponentPipeline,
+    results_getter: ResultsGetter,
+):
+    results_dict = {
```
Comment (lannelin) [inline src/arc_spice/eval/inference_utils.py:101]: why is this here if it gets overwritten?

File: src/arc_spice/eval/inference_utils.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    def __init__(self, n_classes):
+        self.func_map: dict[str, Callable] = {
+            "recognition": self.recognition_results,
+            "translation": self.translation_results,
+            "classification": self.classification_results,
+        }
+        self.comet_model = get_comet_model()
+        self.multihot = MultiHot(n_classes)
+
+    def get_results(
```
Comment (lannelin) [inline src/arc_spice/eval/inference_utils.py:28]: this is actually doing inference if I'm reading this correctly? imo current name suggests it's just fetching precalculated results

File: src/arc_spice/eval/inference_utils.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    ):
+        for step_name in clean_output:
+            results_dict = self.func_map[step_name](
+                test_row, clean_output, var_output, results_dict
+            )
+        results_dict["input_data"]["celex_ids"].append(test_row["celex_id"])
+        return results_dict
+
+    def recognition_results(self, test_row, clean_output, var_output, results_dict):
+        assert test_row is not None
```
Comment (lannelin) [inline src/arc_spice/eval/inference_utils.py:43]: if these are None then the whole process will fail. Should this error be handled instead or are we happy that this should cause program to exit?

File: src/arc_spice/eval/inference_utils.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        assert clean_output is not None
+        assert var_output is not None
+        # ### RECOGNITION ###
+        # TODO: add this into results_getter
```
Comment (lannelin) [inline src/arc_spice/eval/inference_utils.py:47]: ref to issue

File: src/arc_spice/eval/inference_utils.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+from arc_spice.eval.translation_error import get_comet_model
+from arc_spice.variational_pipelines.RTC_single_component_pipeline import (
+    RTCSingleComponentPipeline,
+)
+from arc_spice.variational_pipelines.RTC_variational_pipeline import (
+    RTCVariationalPipeline,
+)
+
+
+class ResultsGetter:
```
Comment (lannelin) [inline src/arc_spice/eval/inference_utils.py:18]: add type hints where missing to methods on this class

File: src/arc_spice/eval/inference_utils.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        # TODO: add this into results_getter
+        return results_dict
+
+    def translation_results(self, test_row, clean_output, var_output, results_dict):
+        # ### TRANSLATION ###
+        source_text = test_row["target_text"]
+        target_text = test_row["target_text"]
+        clean_translation = clean_output["translation"]["full_output"]
+
+        # load error model
```
Comment (lannelin) [inline src/arc_spice/eval/inference_utils.py:56]: does this load the model? it looks already loaded

File: src/arc_spice/eval/inference_utils.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        comet_output = self.comet_model.predict(
+            comet_inp, batch_size=8, accelerator=comet_device
+        )
+        comet_scores = comet_output["scores"]
+        results_dict["translation"]["full_output"].append(clean_translation)
+        results_dict["translation"]["comet_score"].append(comet_scores[0])
+        results_dict["translation"]["weighted_semantic_density"].append(
+            var_output["translation"]["weighted_semantic_density"]
+        )
+        return results_dict
```
Comment (lannelin) [inline src/arc_spice/eval/inference_utils.py:75]: if your results dict has a defined structure then perhaps consider making this a `TranslationResults` class?

Comment (lannelin) [reply src/arc_spice/eval/inference_utils.py:75]: same for other result types

File: src/arc_spice/variational_pipelines/utils.py [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+from typing import Any
+
+import torch
+from torch.nn.functional import softmax
+from transformers import AutoModelForSequenceClassification, AutoTokenizer, Pipeline
+
+logger = logging.Logger("RTC_variational_pipeline")
+
+
+class DummyPipeline:
```
Comment (lannelin) [inline src/arc_spice/variational_pipelines/utils.py:12]: what's the purpose of this class if it will raise NotImplementedError when it's called?

File: src/arc_spice/variational_pipelines/utils.py [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+        training = training_override
+
+    return dropout_orig_fn(input=input, p=p, training=training, inplace=inplace)
+
+
+dropout_on = partial(dropout_w_training_override, training_override=True)
+dropout_off = partial(dropout_w_training_override, training_override=False)
+
+
+class RTCPipelineBase:
```
Comment (lannelin) [inline src/arc_spice/variational_pipelines/utils.py:113]: should probably be an Abstract Base Class as we don't want this to be instantiated? https://docs.python.org/3/library/abc.html
can also define some methods that _have_ to be implemented, if desired.

File: src/arc_spice/variational_pipelines/RTC_single_component_pipeline.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+from transformers import pipeline
+
+from arc_spice.variational_pipelines.RTC_variational_pipeline import (
+    CustomTranslationPipeline,
+    RTCPipelineBase,
+)
+from arc_spice.variational_pipelines.utils import dropout_off, dropout_on, set_dropout
+
+
+class RTCSingleComponentPipeline(RTCPipelineBase):
```
Comment (lannelin) [inline src/arc_spice/variational_pipelines/RTC_single_component_pipeline.py:14]: would this be clearer if split this into 3 separate classes? (possibly keeping shared functionality here as an abstract class)
i.e. TranslationVariationalPipeline, RecognitionVariationalPipeline, TopicVariationalPipeline

File: src/arc_spice/variational_pipelines/utils.py (hunk 11 above)
Comment (lannelin) [inline src/arc_spice/variational_pipelines/utils.py:113]: can we ensure the names for all of these reflect that they're variational? i.e. RTCVariationalPipelineBase

**Review by lannelin — CHANGES_REQUESTED** [review]
nice one. I've made a few small requests for changes and left a few qs. Happy to discuss.

File: config/RTC_configs/bert-mt5-bert.yaml
Comment (J-Dymond) [reply config/RTC_configs/bert-mt5-bert.yaml:1]: Will change to this

File: scripts/pipeline_inference.py (hunk 1 above)
Comment (J-Dymond) [reply scripts/pipeline_inference.py:42]: will do

File: src/arc_spice/eval/inference_utils.py (hunk 8 above)
Comment (J-Dymond) [reply src/arc_spice/eval/inference_utils.py:56]: I can update these comments, I think it used to be loaded here maybe, and I shifted it to the `__init__`.

File: src/arc_spice/eval/inference_utils.py (hunk 5 above)
Comment (lannelin) [reply src/arc_spice/eval/inference_utils.py:43]: I'd prefer not to make the args ambiguous. If this isn't ready yet, should it just raise a NotImplementedError?

File: src/arc_spice/eval/inference_utils.py (hunk 3 above)
Comment (J-Dymond) [reply src/arc_spice/eval/inference_utils.py:101]: It's only overwritten if there is a single component pipeline, I could instead update it to remove the unused rows?

File: src/arc_spice/eval/inference_utils.py (hunk 4 above)
Comment (lannelin) [reply src/arc_spice/eval/inference_utils.py:28]: oh but it calls functions from the func_map and I thought those performed inference?

File: src/arc_spice/variational_pipelines/RTC_single_component_pipeline.py (hunk 12 above)
Comment (J-Dymond) [reply src/arc_spice/variational_pipelines/RTC_single_component_pipeline.py:14]: Sure, is this as subclasses of `RTCSingleComponentPipeline`?

File: src/arc_spice/eval/inference_utils.py (hunk 4 above)
Comment (lannelin) [reply src/arc_spice/eval/inference_utils.py:28]: ah ok because `get_results` *updates* the results_dict. I hadn't realised that. I don't think it's worth changing here but, In future, I think we should avoid having those sort of side effects and where we do need to have them, make them super clear.

File: src/arc_spice/eval/inference_utils.py (hunk 3 above)
Comment (lannelin) [reply src/arc_spice/eval/inference_utils.py:101]: ah ok because get_results updates the results_dict. I hadn't realised that. I don't think it's worth changing here but, In future, I think we should avoid having those sort of side effects and where we do need to have them, make them super clear.

File: src/arc_spice/eval/inference_utils.py (hunk 4 above)
Comment (J-Dymond) [reply src/arc_spice/eval/inference_utils.py:28]: Ah that's my fault, I should name this something else. This is the functions used to obtain results, not the forward methods themselves

File: src/arc_spice/eval/inference_utils.py (hunk 3 above)
Comment (J-Dymond) [reply src/arc_spice/eval/inference_utils.py:101]: Will update the comments to highlight this

Comment (J-Dymond) [reply src/arc_spice/eval/inference_utils.py:114]: I've made changes now, instead returning results as a list of dicts, it uses the `celex_id` row of the dataset as a unique identifier.

File: src/arc_spice/variational_pipelines/utils.py (hunk 10 above)
Comment (J-Dymond) [reply src/arc_spice/variational_pipelines/utils.py:12]: This is for tests so that dummy pipelines can be loaded without the overhead of the models themselves. I also use this class when initialising the pipelines of the base class, this was to ensure mypy didn't raise an attribute error when referencing the pipelines within the class.

File: src/arc_spice/variational_pipelines/utils.py (hunk 11 above)
Comment (J-Dymond) [reply src/arc_spice/variational_pipelines/utils.py:113]: Will look into this

File: src/arc_spice/variational_pipelines/RTC_single_component_pipeline.py (hunk 12 above)
Comment (J-Dymond) [reply src/arc_spice/variational_pipelines/RTC_single_component_pipeline.py:13]: I've implemented this now

File: src/arc_spice/variational_pipelines/utils.py (hunk 10 above)
Comment (lannelin) [reply src/arc_spice/variational_pipelines/utils.py:12]: if it's only for use in tests then can it live in the test code?

Comment (J-Dymond) [reply src/arc_spice/variational_pipelines/utils.py:12]: Sorry, this is still being used in the base class at initialisation. I've changed the error to AttributeError, as this will reflect the nature of the issue if it is called without being properly initialised?

**Review by lannelin — APPROVED** [review]
lgtm.

Could you just add an explanation to the scripts dir README about the new scripts before you merge?

Also, the addition of tests to pre-commit (as you suggested in separate channel) is a good one. Happy for this to be combined in this.
